### PR TITLE
Define `Base.isstored` for Diagonals and Triangular matrices

### LINKED ISSUE #1013: Base.show fails with structural matrices with non-number elements
Some matrices (`SparseMatrixCSC` and `Diagonal`, as far as i know) depend on a type defining `iszero(x::T)` and `zero(Type{T})` to work. a MWE:
```julia
show(IOBuffer(),MIME"text/plain"(),Diagonal(fill(nothing,3)))
```
version info.
```
julia> versioninfo()
Julia Version 1.9.1
Commit 147bdf428c (2023-06-07 08:27 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × Intel(R) Core(TM) i7-4720HQ CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, haswell)
  Threads: 4 on 8 virtual cores
Environment:
  JULIA_NUM_THREADS = 4

```

related:

https://github.com/korsbo/Latexify.jl/issues/269

### PR DESCRIPTION
**Pull request by longemen3000**
Required to fix https://github.com/JuliaLang/LinearAlgebra.jl/issues/1013, moved here from https://github.com/JuliaLang/LinearAlgebra.jl/pull/9

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -162,6 +162,8 @@ end
 diagzero(::Diagonal{T}, i, j) where {T} = zero(T)
 diagzero(D::Diagonal{<:AbstractMatrix{T}}, i, j) where {T} = zeros(T, size(D.diag[i], 1), size(D.diag[j], 2))
 
+Base.isstored(D::Diagonal, i, j) = i == j ? Base.isstored(D.diag,i) : false
```
Comment (jishnub) [inline stdlib/LinearAlgebra/src/diagonal.jl:165]: Perhaps this should be `isstored(::LinearAlgebra.Diagonal, i::Int, j::Int)`

Comment (longemen3000) [reply stdlib/LinearAlgebra/src/diagonal.jl:165]: do you mean typing the `ì` and `j` arguments as `int` ? or putting the method in `Base` instead?

Comment (jishnub) [reply stdlib/LinearAlgebra/src/diagonal.jl:165]: Adding the type assertions, as this will resolve the amgituity that we see in the tests

**Comment by jishnub** [comment]
Should these be added for the other structured matrix types, such as `Bidiagonal`, `Tridiagonal` etc.?

**Comment by longemen3000** [comment]
Those types?
- `UpperTriangular`
- `UnitUpperTriangular`
- `LowerTriangular`
- `UnitLowerTriangular`
- `Tridiagonal`
- `SymTridiagonal`
- `Bidiagonal`

**Comment by jishnub** [comment]
We can start with these, and extend it to others such as `Adjoint` and `Symmetric` later

**Review by vtjnash — APPROVED** [review]

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 2]
```diff
@@ -150,6 +150,16 @@ end
     r
 end
 
+@inline function Base.isstored(D::Diagonal, i::Int, j::Int)
+    @boundscheck checkbounds(A, i, j)
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/diagonal.jl:154]:
```suggestion
    @boundscheck checkbounds(D, i, j)
```

**Pull request merged**